lms/djangoapps/course_blocks/transformers/user_partitions.py

```diff
@@ -79,16 +88,25 @@ def transform_block_filters(self, usage_info, block_structure):
             return [block_structure.create_universal_filter()]
 
         user_groups = get_user_partition_groups(usage_info.course_key, user_partitions, user, 'id')
-        group_access_filter = block_structure.create_removal_filter(
-            lambda block_key: not (
-                has_access(user, 'staff', block_key) or
-                block_structure.get_transformer_block_field(block_key, self, 'merged_group_access').check_group_access(
-                    user_groups
-                )
+        for block_key in block_structure.topological_traversal():
+            transformer_block_field = block_structure.get_transformer_block_field(
+                block_key, self, 'merged_group_access'
             )
-        )
-
-        result_list.append(group_access_filter)
+            access_denying_partition_id = transformer_block_field.get_access_denying_partition(
+                user_groups
+            )
+            access_denying_partition = get_partition_from_id(user_partitions, access_denying_partition_id)
+            if not has_access(user, 'staff', block_key) and access_denying_partition:
+                user_group = user_groups[access_denying_partition_id]
+                allowed_groups = transformer_block_field._access[access_denying_partition_id]
+                access_denied_message = access_denying_partition.access_denied_message(block_key, user, user_group, allowed_groups)
+                block_structure.override_xblock_field(
+                    block_key, 'authorization_denial_reason', access_denying_partition.name
```

Comment:
I guess maybe if there's no message, but there's the partition name, that might be enough data for mobile to do something sensible? It's one of those security/authZ questions of "should we show the user that they don't have access to this, or just hide it entirely?"

Reply:
for now i'm going with your suggestion of hiding it if there is no message